### PR TITLE
[SPARK-7313] [Spark CORE] Configurable limit to range partitioner's max_samples.

### PR DESCRIPTION
**Pull request by mridulm**

### CODE REVIEW

**Comment by mridulm** [comment]
+CC @pwendell We hit this while using non primitive keys.

File: core/src/main/scala/org/apache/spark/Partitioner.scala [hunk 1]
```diff
@@ -118,7 +118,8 @@ class RangePartitioner[K : Ordering : ClassTag, V](
       Array.empty
     } else {
       // This is the sample size we need to have roughly balanced output partitions, capped at 1M.
-      val sampleSize = math.min(20.0 * partitions, 1e6)
+      val maxSamples = rdd.sparkContext.getConf.getDouble("spark.partitioner.max_samples", 1e6)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: `max_samples` -> `maxSamples`. You are trying to turn *down* the max right? Just wondering out loud -- was 1e6 arbitrary and probably too big as a max to begin with? The max would only matter at 50000 partitions, which is huge. Would a max of... 1e5, 1e4 be more sensible for 99% of cases? I realize there's no single value that always works but wondering if we can avoid another flag.

Comment (mridulm) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: Not sure how 1e6 was arrived at, but for our jobs at anything > 40k partitions, for non primitive keys we hit the issue. The actual number of keys is not too large, which is why lower the value does not affect precision of the estimate.

**Comment by mridulm** [comment]
Not added the property to configuration.md - not sure if this is something we want to generally expose. Comments ?

**Comment by srowen** [comment]
Yeah sounds like you need the limit to be < 800k or so. 1 million seems like a really high limit to begin with since it wouldn't even matter for less than 50k partitions and your case is proof that it's not a very effective max. Maybe the max can just be 100k? at least, the default? maybe I'm trying too hard to avoid a flag. I wouldn't document it, no.

**Comment by mridulm** [comment]
100k will definitely fix our immediate concern (we use 500k actually :-) )
But then I am not sure who else might run into the same later.

**Comment by mridulm** [comment]
Any updates on pushing this in ?

File: core/src/main/scala/org/apache/spark/Partitioner.scala [hunk 2]
```diff
@@ -118,7 +118,8 @@ class RangePartitioner[K : Ordering : ClassTag, V](
       Array.empty
     } else {
       // This is the sample size we need to have roughly balanced output partitions, capped at 1M.
-      val sampleSize = math.min(20.0 * partitions, 1e6)
+      val maxSamples = rdd.sparkContext.getConf.getDouble("spark.partitioner.maxSamples", 1e6)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: if anything this should be `spark.rangePartitioner.maxSamples`

**Comment by andrewor14** [comment]
I wonder if we'll ever want `1e6` samples, which seems like a lot. Echoing @srowen's thinking aloud earlier, does it make sense to change this to something like `1e5` instead of making it configurable?

**Comment by mridulm** [comment]
What the exact value should be - I am not sure, the current value does seem to work well for most users - except for degenerate cases like ours : hence hardcoding it leads to problems.
Given this, I would not change the default - rather make it configurable.

**Comment by rxin** [comment]
I'm going to close this pull request. If this is still relevant and you are interested in pushing it forward, please open a new pull request. Thanks!

**Pull request closed without merging**